### PR TITLE
Add some new test cases.

### PR DESCRIPTION
**Pull request by wudwas**
With 5844816, the test errors. Unless I'm missing something, it seems that `substr()` is called once more than required even when `strlen(r) < strlen(s)` should prevent this.

Another problem is that `souffle palindrome1.dl` and `souffle -c palindrome1.dl` give different results:
- `souffle palindrome1.dl`:
```
warning: wrong index position provided by substr("noon",-1,1) functor.
warning: wrong index position provided by substr("hello",-1,1) functor.
warning: wrong index position provided by substr("racecar",-1,1) functor.
warning: wrong index position provided by substr("neveroddoreven",-1,1) functor.
```
- `souffle -c palindrome1.dl`:
```
warning: wrong index position provided by substr("noon",18446744073709551615,1) functor.
warning: wrong index position provided by substr("hello",18446744073709551615,1) functor.
warning: wrong index position provided by substr("racecar",18446744073709551615,1) functor.
warning: wrong index position provided by substr("neveroddoreven",18446744073709551615,1) functor.
```
The compiled C++ first casts to `size_t` and overflows causing the difference in output. So this one is a pretty dodgy test case and can't be added as is, but actually still manages to be correct in the end. I'll delve into these 2 issues more if I can - at this point they appear more like issues on the Souffle end, but thought I'd ask here first for a sanity check. Thanks.

### CODE REVIEW

File: src/RamExecutor.cpp [hunk 1]
```diff
@@ -351,13 +351,34 @@ bool eval(const RamCondition& cond, RamEnvironment& env, const EvalContext& ctxt
                     }
                     return result;
                 }
+                case BinaryConstraintOp::NOT_MATCH: {
+                    RamDomain l = eval(relOp.getLHS(), env, ctxt);
+                    RamDomain r = eval(relOp.getRHS(), env, ctxt);
+                    const std::string& pattern = env.getSymbolTable().resolve(l);
```
Comment (mmcgr) [inline src/RamExecutor.cpp]: This seems like a bad idea - you're making a const string reference to the result of method that returns a const char*. What is actually a reference to?

It does seem to work, though, and is a copy paste of the existing code, so maybe I'm missing something.

Comment (mmcgr) [reply src/RamExecutor.cpp]: So, this is fine. A temporary string is constructed and the const ref can use that. It still seems like a bad idea to me, but it's also valid C++ and the approach we already use.

**Comment by mmcgr** [comment]
I tested Example myself, and oooh, I see what you're talking about with palindrome1 now.

The problem is that the check `strlen(r) < strlen(s)` is carried out after `substr(s, strlen(s)-(strlen(r)+1), 1)`. The invalid starting position is used before the check to see if it is invalid. You will need to find a way to order these operations.

I don't think that this is a bug, other than the compiled version's attempt at `-1`, but the ordering could be more efficient.

For now, you could just remove the palindrome1 test and make an issue with a simplified version.

**Comment by wudwas** [comment]
Thanks, I've removed the `palindrome1` test, so should be good to go now. I also ran checks on the other tests and they are passing.

Yep, your description is correct. I'm not sure if it's a bug either, but since the `strlen(r) < strlen(s)` check fails in the body of the rule, personally I would expect any functions in the head not to be executed at all.

A new issue has been created at #464, so we can continue the discussion there if you'd like.

**Pull request merged**